lib/media/track/localmediatrack.js

```diff
@@ -104,11 +131,60 @@ function mixinLocalMediaTrack(AudioOrVideoTrack) {
 
     stop() {
       this._log.info('Stopping');
+      if (this._workaroundWebKitBug1208516Cleanup) {
+        this._workaroundWebKitBug1208516Cleanup();
+        this._workaroundWebKitBug1208516Cleanup = null;
+      }
       this.mediaStreamTrack.stop();
       this._end();
       return this;
     }
   };
 }
 
+function restartWhenInadvertentlyStopped(localMediaTrack) {
+  const { _getUserMedia: getUserMedia, _log: log } = localMediaTrack;
+
+  function shouldReacquireTrack() {
+    return document.visibilityState === 'visible'
+      && (localMediaTrack.mediaStreamTrack.muted || localMediaTrack.isStopped);
+  }
+
+  function reacquireTrack() {
+    const constraints = { audio: false, video: false };
+    constraints[localMediaTrack.mediaStreamTrack.kind] = localMediaTrack.mediaStreamTrack.getConstraints();
+
+    log.info('Re-acquiring the MediaStreamTrack.');
+    log.debug('Constraints:', constraints);
+    return getUserMedia(constraints).then(mediaStream => {
+      return mediaStream.getTracks()[0];
+    });
+  }
+
+  function handleTrackStateChange() {
+    Promise.race([
+      waitForEvent(localMediaTrack.mediaStreamTrack, 'unmute'),
+      waitForSometime(50)
+    ]).then(() => {
+      if (!shouldReacquireTrack()) {
+        return;
+      }
+      reacquireTrack().then(newTrack => {
+        log.info('Re-acquired the MediaStreamTrack.');
+        log.debug('MediaStreamTrack:', newTrack);
+        localMediaTrack._setMediaStreamTrack(newTrack);
+      }).catch(err => {
+        log.warn('Failed to re-acquire the MediaStreamTrack:', err);
+      });
+    });
+  }
+
+  document.addEventListener('visibilitychange', handleTrackStateChange);
+  localMediaTrack.on('stopped', handleTrackStateChange);
```

Comment:
Wouldn't this risk re-acquiring the track twice?

Reply:
not really, because even if both handlers fire closer to each other `shouldReacquireTrack` would return false for one of them.

Reply:
Thank you for noticing this. - maybe there is a slight chance that they may both endup obtaining track. I will fix this part.